### PR TITLE
Add dispatch mode testing for meta tensors and other stuff

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #77682
* #77681
* #77499
* __->__ #77477

We don't have any coverage for meta tensor correctness for backwards
because torch function mode can only allow us to interpose on
Python torch API calls, but backwards invocations happen from C++.
To make this possible, I add torch_dispatch_meta test which runs the
tests with __torch_dispatch__

While doing this, I needed to generate fresh expected failure / skip
lists for the new test suite, and I discovered that my original
scaffolding for this purpose was woefully insufficient.  So I rewrote
how the test framework worked, and at the same time rewrote the
__torch_function__ code to also use the new logic.  Here's whats
new:

- Expected failure / skip is now done on a per function call basis,
  rather than the entire test.  This means that separate OpInfo
  samples for a function don't affect each other.

- There are now only two lists: expect failure list (where the test
  consistently fails on all runs) and skip list (where the test
  sometimes passes and fails.

- We explicitly notate the dtype that failed.  I considered detecting
  when something failed on all dtypes, but this was complicated and
  listing everything out seemed to be nice and simple.  To keep the
  dtypes short, I introduce a shorthand notation for dtypes.

- Conversion to meta tensors is factored into its own class
  MetaConverter

- To regenerate the expected failure / skip lists, just run with
  PYTORCH_COLLECT_EXPECT and filter on a specific test type
  (test_meta or test_dispatch_meta) for whichever you want to update.

Other misc fixes:

- Fix max_pool1d to work with BFloat16 in all circumstances, by making
  it dispatch and then fixing a minor compile error (constexpr doesn't
  work with BFloat16)

- Add resolve_name for turning random torch API functions into string
  names

- Add push classmethod to the Mode classes, so that you can more easily
  push a mode onto the mode stack

- Add some more skips for missing LAPACK

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

**Comment by zou3519** [comment]
> To regenerate the expected failure / skip lists, just run with
PYTORCH_COLLECT_EXPECT and filter on a specific test type
(test_meta or test_dispatch_meta) for whichever you want to update.

Oh I want this for functorch

File: torch/utils/_python_dispatch.py [hunk 1]
```diff
@@ -147,6 +147,10 @@ def __init__(self):
     def __torch_dispatch__(self, func, types, args=(), kwargs=None):
         raise NotImplementedError()
 
+    @classmethod
+    def push(cls, *args, **kwargs):
+        return push_torch_dispatch_mode(functools.partial(cls, *args, **kwargs))
```
Comment (zou3519) [inline torch/utils/_python_dispatch.py:152]: Add test in test_python_dispatch?

**Comment by ezyang** [comment]
> very cool, I want this for functorch

Unfortunately, I basically completely bypassed the OpInfo xfail mechanism and have my own. I think from first principles this makes sense anyway, because OpInfos can only operate at the unit test granularity, but here I actually know if any given operation *inside* the script is expected to fail or not.

File: torch/utils/_python_dispatch.py (hunk 1 above)
Comment (ezyang) [reply torch/utils/_python_dispatch.py:152]: Actually, I was thinking of gussying up the API even more. What if this worked:

```
with MyMode(blah, blah):
  ...
```

The main missing ingredient is that the mode constructor needs to record the current mode to its `inner` (instead of getting it mutated by the push function, which is currently what's happening).

File: torch/overrides.py [hunk 2]
```diff
@@ -1554,36 +1555,32 @@ def handle_torch_function(
...
+        ("torch.special", torch.special, dir(torch.special)),
     ]
-    for namespace, ns_funcs in tested_namespaces:
+    for namespace_str, namespace, ns_funcs in tested_namespaces:
         for func_name in ns_funcs:
+            ignore = False
             # ignore private functions or functions that are deleted in torch.__init__
             if namespace is not torch.Tensor:
-                if func_name.startswith('_'):
+                if func_name.startswith('__'):
```
Comment (zou3519) [inline torch/overrides.py:1576]: This wasn't here before, does it matter?

File: torch/csrc/utils/python_dispatch.cpp [hunk 3]
```diff
@@ -202,6 +202,15 @@ void initDispatchBindings(PyObject* module) {
     c10::Dispatcher::singleton().checkInvariants();
   });
 
+  m.def("_dispatch_has_kernel_for_dispatch_key", [](const char* name, std::string dispatch) -> bool {
+    auto op = c10::Dispatcher::singleton().findOp(torch::jit::parseName(name));
+    if (!op) {
+      return "";
```
Comment (zou3519) [inline torch/csrc/utils/python_dispatch.cpp:208]: is this a bool?

File: torch/csrc/utils/python_dispatch.cpp [hunk 4]
```diff
@@ -202,6 +202,15 @@ void initDispatchBindings(PyObject* module) {
     c10::Dispatcher::singleton().checkInvariants();
   });
 
+  m.def("_dispatch_has_kernel_for_dispatch_key", [](const char* name, std::string dispatch) -> bool {
```
Comment (zou3519) [inline torch/csrc/utils/python_dispatch.cpp:205]: Does `const std::string&` work for both args?

File: torch/overrides.py [hunk 5]
```diff
@@ -1615,14 +1616,49 @@ def get_overridable_functions() -> Dict[Any, List[Callable]]:
...
+
+    Returns
+    -------
+    Dict[Any, List[Callable]]
+        A dictionary that maps namespaces that contain overridable functions
+        to functions in that namespace that can be overridden.
+    """
+    return _get_overridable_functions()[0]
+
+def resolve_name(f):
```
Comment (zou3519) [inline torch/overrides.py:1644]: unit test would be nice for this

File: torch/_decomp/__init__.py [hunk 6]
```diff
@@ -15,7 +15,7 @@
 meta_lib = torch.library.Library("aten", "IMPL", "Meta")
 
 
-def register_decomposition(aten_op, registry=None, *, register_meta: bool = False):
+def register_decomposition(aten_op, registry=None, *, disable_meta: bool = False):
```
Comment (zou3519) [inline torch/_decomp/__init__.py:18]: (no action required) ack, people are going to add disable_* flags for every subsystem as they come along. Not sure if that's the UX we want

File: torch/_decomp/__init__.py [hunk 7]
```diff
@@ -53,7 +53,11 @@ def add_op_to_table(aten_op):
                 if op_overload in registry:
                     raise RuntimeError(f"duplicate registrations for {op_overload}")
                 registry[op_overload] = f
-                if register_meta:
+                # TODO: factor this logic into OpOverload or Library API
+                name = op_overload._schema.name
+                if op_overload._schema.overload_name:
+                    name += "." + op_overload._schema.overload_name
```
Comment (zou3519) [inline torch/_decomp/__init__.py:59]: nit: FunctionSchema's pybind binding should really have a .operator_name() like the C++ FunctionSchema so we don't need to do this

File: test/test_meta.py [hunk 8]
```diff
@@ -801,30 +977,32 @@ class TestMeta(TestCase):
...
+    def test_dispatch_meta(self, device, dtype, op):
+        func = op.get_op()
+        samples = op.sample_inputs(device, dtype, requires_grad=False)
+        for sample_input in samples:
+            args = [sample_input.input] + list(sample_input.args)
+            kwargs = sample_input.kwargs
+            with MetaCrossRefDispatchMode.push(self, dtype=dtype, device=device):
+                expected = func(*args, **kwargs)
+                if isinstance(expected, torch.Tensor) and op.supports_out:
+                    func(*args, **kwargs, out=expected)
```
Comment (zou3519) [inline test/test_meta.py:1005]: Did you want to test backward? Or is that coming later?

File: test/test_meta.py [hunk 9]
```diff
@@ -668,129 +117,856 @@ def meta_tensor(t):
...
+}
+
+"""
+import yaml
+print(yaml.dump({"boof":
+  {resolve_name(k): [dtype_abbrs[d] for d in v]
+   for k, v in meta_function_expected_failures.items()}}, default_flow_style=None))
+import sys
+sys.exit()
+"""
```
Comment (zou3519) [inline test/test_meta.py:477]: Is this meant to be here? If so, add a comment?

**Review by zou3519 — COMMENTED** [review]
Need to run, I got through everything except for the xfail generation and the refs file (don't think I am qualified to review that). Will resume later.

High level comment is that this looks good but is a bit difficult to review, would probably have been better to split it up more (but I'm not sure how, maybe the problem is just that the blame makes it hard to tell what got removed and what didn't)? Also some minor nits.

File: torch/utils/_python_dispatch.py (hunk 1 above)
Comment (zou3519) [reply torch/utils/_python_dispatch.py:152]: Yeah this would be cool!

File: torch/overrides.py (hunk 2 above)
Comment (ezyang) [reply torch/overrides.py:1576]: Yeah, without this I pick up stuff like `__name__` from the module that I don't want. We only have magic methods on Tensor.

File: torch/csrc/utils/python_dispatch.cpp (hunk 3 above)
Comment (ezyang) [reply torch/csrc/utils/python_dispatch.cpp:208]: whoops

File: torch/csrc/utils/python_dispatch.cpp (hunk 4 above)
Comment (ezyang) [reply torch/csrc/utils/python_dispatch.cpp:205]: i'll make them both const char*

File: test/test_meta.py (hunk 8 above)
Comment (ezyang) [reply test/test_meta.py:1005]: coming later, i'll have to redo the skips when I do this

File: torch/utils/_python_dispatch.py (hunk 1 above)
Comment (ezyang) [reply torch/utils/_python_dispatch.py:152]: cc @samdow 

What I'm vacillating about is how exactly this should be implemented. An obvious way to do it is

```
class ModeBase:
    def __init__(self):
        self.inner = _get_mode()
```

but I'm afraid people will forget to call `super().__init__` and then inner won't be set (maybe we can detect this happened with hasattr). Alternately, we can use wrap_init in the metaclass and that also makes easy to accept an `inner` kwarg to override it, but it's more implicit.

Comment (samdow) [reply torch/utils/_python_dispatch.py:152]: > I'm afraid people will forget to call super().__init__ and then inner won't be set (maybe we can detect this happened with hasattr)

This should be caught during the first call to torch dispatch. We could add a clearer error message with the hasattr check since we have a guess as to what the user did wrong (in general, they shouldn't have to think about inner that much)

> Alternately, we can use wrap_init in the metaclass and that also makes easy to accept an inner kwarg to override it, but it's more implicit.

IMO, this feel nicer since my guess is that most people will want the implicit behavior of just using whatever mode is currently inner. Also since we're already using the metaclass to wrap torch dispatch and init, it fits with the rest of the design. I guess the only question is the flip of the current question: in this version, what happens if a user calls `super().__init__`? Does wrap_init get called twice (maybe not the end of the world)?

File: torch/_refs/__init__.py [hunk 10]
```diff
@@ -411,7 +410,7 @@ def _ref(
     if aten_op is infer_aten_op:
         aten_op = getattr(torch.ops.aten, prim.__name__.split(".")[0])
     if aten_op is not None:
-        register_decomposition(aten_op, register_meta=register_meta)(_ref)
+        register_decomposition(aten_op, disable_meta=disable_meta)(_ref)
```
Comment (zou3519) [inline torch/_refs/__init__.py:413]: primtorch refs go into the decomposition table too? That's interesting -- are there cases where we have multiple decompositions registered for an op?

**Review by zou3519 — APPROVED** [review]
Okay, got through the files. Some minor comments/questions but otherwise lgtm

File: test/test_meta.py [hunk 11]
```diff
@@ -668,129 +117,858 @@ def meta_tensor(t):
...
+    # For now, only attempt if we managed to convert all tensor types
+    # (if any of them failed, we're in a mixed device situation and
+    # this isn't well supported)
+    if do_meta and to_meta.successful():
+        try:
+            # Suppress warnings, this matters because some tests are
+            # checking warnings!
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore")
+                meta_rs = func(*meta_args, **meta_kwargs)
```
[the remark this answers is not in the document]
Comment (ezyang) [reply test/test_meta.py:291]: Oh yeah, this is a holdover from when I was using this logic to do crossref testing.

File: test/test_meta.py [hunk 12]
```diff
@@ -668,129 +117,858 @@ def meta_tensor(t):
...
+            meta_r.storage_offset() == r.storage_offset(),
+            f"but real storage_offset was {r.storage_offset()}")
+        test_assert(meta_r.requires_grad == r.requires_grad, f"but real requires_grad was {r.requires_grad}")
+        test_assert(meta_r.is_conj() == r.is_conj(), f"but real is_conj was {r.is_conj()}")
+        test_assert(meta_r.is_neg() == r.is_neg(), f"but real is_neg was {r.is_neg()}")
 
-        rs = func(*args, **kwargs)
 
-        # TODO: also handle cases where func raise an exception
+COLLECT_EXPECT = os.getenv('PYTORCH_COLLECT_EXPECT', '0') == '1'
```
[the remark this answers is not in the document]
Comment (ezyang) [reply test/test_meta.py:195]: Yes, the insertion is still manual. Not sure if it's worth investing in making it more automatic.

File: torch/_refs/__init__.py (hunk 10 above)
Comment (zou3519) [reply torch/_refs/__init__.py:413]: I'd expect the primtorch decomp to be a lot slower due to needing to model things like type promotion, does that matter in practice?

File: torch/utils/_python_dispatch.py (hunk 1 above)
Comment (ezyang) [reply torch/utils/_python_dispatch.py:152]: If we go the meta class route I would just not put anything in the super init. I guess there's a weird problem which is that the metaclass will try to set inner on every super() call, that's kind of weird, not sure if there's a good way around it

File: torch/overrides.py (hunk 2 above)
Comment (zou3519) [reply torch/overrides.py:1578]: sgtm, thanks for clarifying

File: torch/csrc/utils/python_dispatch.cpp [hunk 13]
```diff
@@ -202,6 +202,15 @@ void initDispatchBindings(PyObject* module) {
     c10::Dispatcher::singleton().checkInvariants();
   });
 
+  m.def("_dispatch_has_kernel_for_dispatch_key", [](const char* name, const char* dispatch) -> bool {
+    auto op = c10::Dispatcher::singleton().findOp(torch::jit::parseName(name));
+    if (!op) {
+      return false;  // TODO: or raise an error?
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/csrc/utils/python_dispatch.cpp]: So it turns out this is very annoying, because TorchScript adds a bunch of extra overloads to operators that aren't actually in the dispatcher.  I ended up with this in the end:

```
                if (
                    not disable_meta
                    # TorchScript dumps a bunch of extra nonsense overloads
                    # which don't have corresponding dispatcher entries, we need
                    # to filter those out
                    and torch._C._dispatch_has_kernel(name)
                    and not torch._C._dispatch_has_kernel_for_dispatch_key(name, 'Meta')
                ):  
                    meta_lib.impl(op_overload, f)
```

File: test/test_meta.py [hunk 14]
```diff
@@ -1,657 +1,107 @@
...
-        miss = 0
+# This is a class for converting multiple tensors into meta tensors which
+# share the same view/storage structure.  The operation model is you allocate
+# one of these, and then call it repeatedly on all the tensors you want to
+# convert.  It's important to use the same object for tensors you want to
+# share storage because this is how we correlate shared storages to the same
+# meta storages; similarly, it's important NOT to use the same object for
+# unrelated groups of tensors because this class will remember all the
+# tensors/storages its seen and therefore leak memory.
+class MetaConverter:
```
Comment (eellison) [inline test/test_meta.py:76]: is this worth factoring out outside of test? didn't we have a `#TODO: aliasing` somewhere previously ?

File: test/test_meta.py [hunk 15]
```diff
@@ -1,657 +1,107 @@
...
             with torch.inference_mode(t.is_inference()):
-                s = meta_storage(t.storage())
+                s = self.meta_storage(t.storage())
                 is_leaf = safe_is_leaf(t)
                 if is_leaf or not t._is_view():
                     r = torch.empty(
                         (0,), dtype=t.dtype, device='meta'
                     )
-                    r.set_(s, t.storage_offset(), t.size(), t.stride())
+                    with no_dispatch():
```
Comment (eellison) [inline test/test_meta.py:103]: just for my own edification - why is this necessary ?

File: test/test_meta.py [hunk 16]
```diff
@@ -668,129 +118,875 @@ def meta_tensor(t):
...
-            except Exception as e:
-                raise RuntimeError(
-                    f"failed to convert args to meta; "
-                    f"originally (*{args}, **{kwargs})") from e
+def assert_ref_meta_equal(test_case, meta_rs, rs, msg_callable):
+    flat_meta_rs, _ = tree_flatten(meta_rs)
+    flat_rs, _ = tree_flatten(rs)
+    test_case.assertEqual(len(flat_meta_rs), len(flat_rs))
+    for i, meta_r, r in zip(range(len(flat_rs)), flat_meta_rs, flat_rs):
+        def test_assert(cond, msg):
```
Comment (eellison) [inline test/test_meta.py:179]: is it worth asserting on the shape of the tree here ? I've seen instances in past where say a meta decomposition returned a list but the corresponding function returned a tuple

File: test/test_meta.py [hunk 17]
```diff
@@ -668,129 +118,875 @@ def meta_tensor(t):
...
-        rs = func(*args, **kwargs)
+# This environment variable controls whether or not we print expected failure
+# lists at the end of a test suite run.  The intended usage looks like this:
+#
+# 1. Run `PYTORCH_COLLECT_EXPECT=1 python test/test_meta.py` on a CUDA build
+#    of PyTorch that has LAPACK/MAGMA installed.  You can filter `-k test_meta`
+#    or `-k test_dispatch_meta` to only focus on one or another list
+# 2. Given the printed skip/xfail list, add them to the corresponding lists;
+#    torch.* entries go in meta_function and aten.* entries go in meta_dispatch.
+#    If there are preexisting entries, you need to merge in the entries.
```
Comment (eellison) [inline test/test_meta.py:204]: 😍 😍 😍

File: test/test_meta.py [hunk 18]
```diff
@@ -801,30 +997,32 @@ class TestMeta(TestCase):
     def test_meta(self, device, dtype, op):
         # run the OpInfo sample inputs, cross-referencing them with the
         # meta implementation and check the results are the same.  All
-        # the heavy lifting happens in MetaCrossRefMode
+        # the heavy lifting happens in MetaCrossRefFunctionMode
         func = op.get_op()
```
Comment (eellison) [inline test/test_meta.py:1001]: is it worth using the api to get more samples ?

File: torch/overrides.py [hunk 19]
```diff
@@ -1556,36 +1557,32 @@ def handle_torch_function(
...
-        (torch.fft, dir(torch.fft)),
-        (torch.special, dir(torch.special)),
+        ("torch", torch, torch.__all__ + dir(torch._C._VariableFunctions)),
+        ("torch.functional", torch.functional, torch.functional.__all__),
+        ("torch.nn.functional", torch.nn.functional, dir(torch.nn.functional)),
+        ("torch.nn.init", torch.nn.init, dir(torch.nn.init)),
+        ("torch.Tensor", torch.Tensor, dir(torch.Tensor)),
+        ("torch.linalg", torch.linalg, dir(torch.linalg)),
+        ("torch.fft", torch.fft, dir(torch.fft)),
+        ("torch.special", torch.special, dir(torch.special)),
```
Comment (eellison) [inline torch/overrides.py:1571]: now that these are overrideable... should we have bc testing like we do with operators ?

**Review by eellison — COMMENTED** [review]
cool!

File: test/test_meta.py (hunk 14 above)
Comment (ezyang) [reply test/test_meta.py:76]: Yeah probably! Since it's top level now easy enough to move

File: test/test_meta.py (hunk 15 above)
Comment (ezyang) [reply test/test_meta.py:103]: It's very dumb... set_ with Tensor argument redispatches to set_ with Storage which would call back into Python when there's a mode... but we can't because meta storages don't work in Python lol

File: test/test_meta.py (hunk 16 above)
Comment (ezyang) [reply test/test_meta.py:179]: Yes I agree

File: test/test_meta.py [hunk 20]
```diff
@@ -668,129 +118,875 @@ def meta_tensor(t):
...
+    flat_meta_rs, _ = tree_flatten(meta_rs)
+    flat_rs, _ = tree_flatten(rs)
+    test_case.assertEqual(len(flat_meta_rs), len(flat_rs))
+    for i, meta_r, r in zip(range(len(flat_rs)), flat_meta_rs, flat_rs):
+        def test_assert(cond, msg):
+            if not cond:
+                raise RuntimeError(f"output {i}: {msg_callable(msg)}")
+        if not isinstance(r, torch.Tensor):
+            continue
+        test_assert(isinstance(meta_r, torch.Tensor), f"but real {i}th result is Tensor")
```
[the remark this answers is not in the document]
Comment (ezyang) [reply test/test_meta.py:184]: This seems less useful, since typically all the out tensors are fresh unless this is specifically a view

File: test/test_meta.py (hunk 18 above)
Comment (ezyang) [reply test/test_meta.py:1001]: I actually wasn't aware how to get more samples, but yes, seems useful?

File: torch/csrc/utils/python_dispatch.cpp [hunk 21]
```diff
@@ -202,6 +202,17 @@ void initDispatchBindings(PyObject* module) {
     c10::Dispatcher::singleton().checkInvariants();
   });
 
+  m.def("_dispatch_has_kernel", [](const char* name) -> bool {
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/csrc/utils/python_dispatch.cpp:205]: Yes, though name is consistent with the rest of the naming in these bindings

File: test/test_meta.py (hunk 18 above)
Comment (eellison) [reply test/test_meta.py:1001]: https://github.com/pytorch/pytorch/blob/bdbb7fe37a49cb034773995a4ec192ba5b1b7ba8/torch/testing/_internal/common_methods_invocations.py#L409

https://github.com/pytorch/pytorch/blob/bdbb7fe37a49cb034773995a4ec192ba5b1b7ba8/test/test_ops.py#L313

File: torch/overrides.py (hunk 19 above)
Comment (eellison) [reply torch/overrides.py:1571]: i'm not saying we're not allowed, but we probably dont want to do so accidentally either

File: test/test_meta.py [hunk 22]
```diff
@@ -668,129 +118,875 @@ def meta_tensor(t):
                         r = base.as_strided(t.size(), t.stride(), t.storage_offset())
```
[the remark this answers is not in the document]
Comment (ezyang) [reply test/test_meta.py:118]: Oops you are right, this is wrong

File: torch/utils/_python_dispatch.py (hunk 1 above)
Comment (samdow) [reply torch/utils/_python_dispatch.py:152]: I agree setting inner on every super call is kind of weird and we could do something like with the hasattr if we really wanted to only set it once. I still think it's better than having an error that makes the user have to know about inner even in the basic case

File: torch/csrc/utils/python_dispatch.cpp (hunk 21 above)
Comment (anjali411) [inline torch/csrc/utils/python_dispatch.cpp:205]: nice this is exactly what I needed to disallow access to overloads added by jit

File: torch/_decomp/__init__.py [hunk 23]
```diff
@@ -53,7 +53,18 @@ def add_op_to_table(aten_op):
...
+                if (
+                    not disable_meta
+                    # TorchScript dumps a bunch of extra nonsense overloads
+                    # which don't have corresponding dispatcher entries, we need
+                    # to filter those out
+                    and torch._C._dispatch_has_kernel(name)
```
Comment (anjali411) [inline torch/_decomp/__init__.py:65]: it would be nicer to add this filter to the `overloads()` method on `OpOverloadPacket since we wanted to disallow access to jit overloads anyway

File: torch/_decomp/__init__.py [hunk 24]
```diff
@@ -53,7 +53,18 @@ def add_op_to_table(aten_op):
                 if op_overload in registry:
                     raise RuntimeError(f"duplicate registrations for {op_overload}")
```
Comment (anjali411) [inline torch/_decomp/__init__.py:54]: could we change the above to `torch.ops.OpOverloadPacket` and `torch.ops.OpOverload`

File: torch/overrides.py [hunk 25]
```diff
@@ -1556,36 +1557,32 @@ def handle_torch_function(
...
         for func_name in ns_funcs:
+            ignore = False
             # ignore private functions or functions that are deleted in torch.__init__
             if namespace is not torch.Tensor:
-                if func_name.startswith('_'):
+                if func_name.startswith('__'):
                     continue
+                elif func_name.startswith('_'):
+                    ignore = True
                 elif func_name.endswith('_'):
```
Comment (anjali411) [inline torch/overrides.py:1582]: why are these functions ignored?

File: torch/_decomp/__init__.py (hunk 7 above)
Comment (ezyang) [reply torch/_decomp/__init__.py:59]: There's an annoying API mismatch, where some APIs require `foo.default` and others require `foo` (no appended dot).

File: test/test_meta.py (hunk 18 above)
Comment (ezyang) [reply test/test_meta.py:1001]: Will do it later

File: torch/overrides.py (hunk 19 above)
Comment (ezyang) [reply torch/overrides.py:1571]: filed here https://github.com/pytorch/pytorch/issues/77668

File: torch/csrc/utils/python_dispatch.cpp (hunk 21 above)
Comment (ezyang) [reply torch/csrc/utils/python_dispatch.cpp:205]: @anjali411 Do you think we can do some more general cleanup here? Maybe version `overloads()` that only returns dispatcher overloads?

File: torch/_decomp/__init__.py (hunk 24 above)
Comment (ezyang) [reply torch/_decomp/__init__.py:54]: but that spelling doesn't work? 

```
>>> torch._ops.OpOverloadPacket
<class 'torch._ops.OpOverloadPacket'>
>>> torch.ops.OpOverloadPacket
<module 'torch.ops.OpOverloadPacket' from 'torch.ops'>
```

File: torch/_refs/__init__.py (hunk 10 above)
Comment (zou3519) [reply torch/_refs/__init__.py:413]: > It seems better to trade performance for correctness

I agree that it's better to trade performance for correctness.

File: torch/csrc/utils/python_dispatch.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/csrc/utils/python_dispatch.cpp]: Yikes, thanks for the explanation

File: torch/_decomp/__init__.py (hunk 23 above)
Comment (ezyang) [reply torch/_decomp/__init__.py:65]: It's not obvious to me that this is the right thing to do. In particular, these overloads DO participate in torch.ops.aten.foo resolution, so it seems weird to filter them out from the API here.

**Comment by ezyang** [comment]
@eellison it turns out MetaConverter was buggy (in ways that don't matter for OpInfo testing), but I had to fix it so it's fixed now and has its own little test suite too.

File: torch/_decomp/__init__.py (hunk 24 above)
Comment (anjali411) [reply torch/_decomp/__init__.py:54]: ugh okay. I am not sure what to do but it seems like these classes should be accessible from a namespace not beginning with an underscore since they are pretty much a public API at this point.

Comment (ezyang) [reply torch/_decomp/__init__.py:54]: Another fix is to put them in another module, not torch._ops

File: torch/overrides.py (hunk 25 above)
Comment (anjali411) [reply torch/overrides.py:1582]: ah I was referring to this: `func_name.endswith('_')`. `func_name.startswith('_')` is understandably addressing the private API

File: test/test_meta.py [hunk 26]
```diff
@@ -1,607 +1,760 @@
...
+        # inference mode can trigger this
+        return False
+
+
+# This is a class for converting multiple tensors into meta tensors which
+# share the same view/storage structure.  The operation model is you allocate
+# one of these, and then call it repeatedly on all the tensors you want to
+# convert.  It's important to use the same object for tensors you want to
+# share storage because this is how we correlate shared storages to the same
+# meta storages; similarly, it's important NOT to use the same object for
```
Comment (eellison) [inline test/test_meta.py:74]: Can we make this use `weakrefs` instead ?
